harness/determined/_preemption.py

```diff
@@ -0,0 +1,142 @@
+import json
+import selectors
+import socket
+import threading
+from typing import Any, Optional
+
+from determined.common import api
+from determined.common.api import certs
+
+
+class _PreemptionWatcher(threading.Thread):
+    """
+    _PreemptionWatcher connects to the master and asynchronously waits for a preemption signal.
+
+    _PreemptionWatcher.should_preempt() is non-blocking (after the initial contact is made with the
+    master) and returns a bool indicating if a preemption signal has been received yet.
+
+    Example usage:
+
+    .. code:: python
+
+       with _PreemptionWatcher(master_url, trial_id, cert) as p:
+           print("started!")
+           for i in range(10):
+               if p.should_preempt():
+                   print("preempted!")
+                   break
+               print('no preemption yet, waiting...')
+               time.sleep(1)
+           else:
+               print('finished without preemption signal')
+    """
+
+    def __init__(self, master_url: str, trial_id: int, cert: certs.Cert) -> None:
+        self._master_url = master_url
+        self._trial_id = trial_id
+        self._cert = cert
+
+        # We will use a selector loop as a non-blocking way to stream through the requests library.
+        # We use a socketpair for a cross-platform way to wake up a selector loop mid-stream.
+        self._rsock = None  # type: Optional[socket.socket]
+        self._wsock = None  # type: Optional[socket.socket]
+
+        self._should_preempt = None  # type: Optional[bool]
+
+        self._cond = threading.Condition()
+        super().__init__()
+
+    def run(self) -> None:
+        assert self._rsock is not None
+        assert self._wsock is not None
+
+        with api.get(
+            self._master_url,
+            f"/api/v1/trials/{self._trial_id}/signals/preemption",
+            stream=True,
+            cert=self._cert,
+        ) as r:
+            lines = r.iter_lines()
+
+            # The first response is always immediate, so we always block for it.  This is doubly
+            # important because we can only select on the HTTP connection socket _after_ we call
+            # next(lines) at least once.
+            for line in lines:
```

Comment:
let's say 
- there's still no first response, so we are blocking here.
- `exit()` is called. it'd stop/return this thread after line 86, but if we are stuck here, it'd be stuck "forever"/until the network timeout.

perhaps all the reading in this method should only be done after the selectors are registered.

Reply:
A couple things here:

* if `exit()` was called on the main thread, you'd only have to block until the first round-trip from the master was complete, because the master _must_ send the first response immediately.  That would be pretty trivial in practice, unless the master was not running and there was a firewall dropping ports.  In that case, blocking until a network timeout seems quite reasonable to me.
* it would be nice to not block here, even for one roundtrip to the master, but there's a practical consideration here which is that `select` does not work on the `r.raw` object until after streaming the first request from the master.  I didn't delve deep enough into urllib3 to try to figure out why that was.

Reply:
Ok, after poking around a bit, it seems highly likely that the reason `select` is not working for before the first response is because the first response out of `iter_lines` arrives in the same network packet as the header of the HTTP response, and has already been recv'd off of the socket.

We've subclassed Requests before to make that happen.  Maybe there is a way here as well.

Reply:
I dug even deeper into this.

* a `requests.Response` object contains a `urllib3.HTTPResponse` object
* a `urllib3.HTTPResponse` object contains an `http.client.HTTPResponse` object
* the `http.client.HTTPResponse` object contains an `io.BufferedRWPair`-wrapped socket object.
* the `BufferedRWPair` can consume data off of the wire in an opaque way, making it unreliable to use `select` to decide if there `read()` will block
* PEP 3116 indicates that it's ok to just make the underlying filedescriptor non-blocking, in which case the `BufferedRWPair` [will raise `EAGAIN` errors](https://www.python.org/dev/peps/pep-3116/#non-blocking-i-o)
* If we did that, we can basically count on at least one of the `requests`, `urllib3`, or `http.client` objects becoming unusable due to the libraries assuming IO exceptions are non-recoverable.

TL;DR I'm convinced that using `select` like this is an absolutely terrible idea.

P.S: "doesn't this code work as-written?  Why wouldn't we just use this?"  No, this code would hang forever if the master ever sent the second response so quickly that both streamed responses appear in a single `socket.recv()` call from the operating system.